### PR TITLE
[WIP] feat: use sqlite as lock for snap management

### PR DESCRIPTION
**Pull request by IronCore864**
Proof of concept, _do not merge_.

This PR uses SQLite as a lock and reference counter when handling snaps. Each unit "registers" itself with a snap when installing and "unregisters" itself from that snap when uninstalling. The snap would only be removed if there are no other units registered with the snap.

See manual tests below.

Todo:

- Handle corner cases with defensive programming. For example, when a unit is already registered with the snap.
- Design the class/interface, maybe merge it with `snap` within `operator-libs-linux`, or make it a standalone class.
- Merge configs.

### CODE REVIEW

**Comment by IronCore864** [comment]
Deploy the first principal charm, the app is named `microsample-vm1`:

```bash
$ juju deploy ./microsample-vm_ubuntu-22.04-arm64.charm microsample-vm1
$ juju status
Model      Controller  Cloud/Region         Version  SLA          Timestamp
lxd-model  lxd         localhost/localhost  3.6.5    unsupported  09:33:48+08:00

App              Version  Status  Scale  Charm           Channel  Rev  Exposed  Message
microsample-vm1           active      1  microsample-vm             9  no

Unit                Workload  Agent  Machine  Public address  Ports  Message
microsample-vm1/1*  active    idle   4        10.15.210.174

Machine  State    Address        Inst id        Base          AZ  Message
4        started  10.15.210.174  juju-8eb266-4  ubuntu@22.04      Running
```

Deploy the second principal charm, the app is named `microsample-vm2`, to the same machine:

```bash
$ juju deploy --to 4 ./microsample-vm_ubuntu-22.04-arm64.charm microsample-vm2
$ juju status
Model      Controller  Cloud/Region         Version  SLA          Timestamp
lxd-model  lxd         localhost/localhost  3.6.5    unsupported  09:34:15+08:00

App              Version  Status  Scale  Charm           Channel  Rev  Exposed  Message
microsample-vm1           active      1  microsample-vm             9  no
microsample-vm2           active      1  microsample-vm            10  no

Unit                Workload  Agent  Machine  Public address  Ports  Message
microsample-vm1/1*  active    idle   4        10.15.210.174
microsample-vm2/2*  active    idle   4        10.15.210.174

Machine  State    Address        Inst id        Base          AZ  Message
4        started  10.15.210.174  juju-8eb266-4  ubuntu@22.04      Running
```

Deploy the `opentelemetry-collector-operator` charm, which manages both `opentelemetry-collector` (`otelcol`) and `node-exporter` snaps/binaries:

```bash
$ juju deploy ./opentelemetry-collector_ubuntu@22.04-arm64.charm
$ juju status
Model      Controller  Cloud/Region         Version  SLA          Timestamp
lxd-model  lxd         localhost/localhost  3.6.5    unsupported  09:36:06+08:00

App                      Version  Status   Scale  Charm                    Channel  Rev  Exposed  Message
microsample-vm1                   active       1  microsample-vm                      9  no
microsample-vm2                   active       1  microsample-vm                     10  no
opentelemetry-collector           unknown      0  opentelemetry-collector            16  no

Unit                Workload  Agent  Machine  Public address  Ports  Message
microsample-vm1/1*  active    idle   4        10.15.210.174
microsample-vm2/2*  active    idle   4        10.15.210.174

Machine  State    Address        Inst id        Base          AZ  Message
4        started  10.15.210.174  juju-8eb266-4  ubuntu@22.04      Running
```

Since `opentelemetry-collector-operator` is a subordinate charm, after it's deployed, nothing happens, there is no unit for it. If we SSH into the machine, we can verify the snaps aren't installed yet:

```bash
$ juju ssh 10.15.210.174
ubuntu@juju-8eb266-4:~$ ls /snap/
README  bin  core  core20  lxd  microsample  snapd
```

Now we add the first subordinate relation:

```bash
$ juju integrate microsample-vm1 opentelemetry-collector
$ juju status
Model      Controller  Cloud/Region         Version  SLA          Timestamp
lxd-model  lxd         localhost/localhost  3.6.5    unsupported  09:38:33+08:00

App                      Version  Status  Scale  Charm                    Channel  Rev  Exposed  Message
microsample-vm1                   active      1  microsample-vm                      9  no
microsample-vm2                   active      1  microsample-vm                     10  no
opentelemetry-collector           active      1  opentelemetry-collector            16  no

Unit                           Workload  Agent  Machine  Public address  Ports  Message
microsample-vm1/1*             active    idle   4        10.15.210.174
  opentelemetry-collector/32*  active    idle            10.15.210.174
microsample-vm2/2*             active    idle   4        10.15.210.174

Machine  State    Address        Inst id        Base          AZ  Message
4        started  10.15.210.174  juju-8eb266-4  ubuntu@22.04      Running
```

In the machine:

```bash
ubuntu@juju-8eb266-4:~$ ls /snap/
README  bin  core  core20  core24  lxd  microsample  node-exporter  opentelemetry-collector  snapd
```

The snaps are installed.

In `juju debug-log`

```bash
unit-opentelemetry-collector-32: 09:38:12 DEBUG unit.opentelemetry-collector/32.juju-log ======= Unit name: opentelemetry-collector/32, snap: opentelemetry-collector registered
unit-opentelemetry-collector-32: 09:38:12 DEBUG unit.opentelemetry-collector/32.juju-log ======= Unit name: opentelemetry-collector/32, snap: opentelemetry-collector used by: ['opentelemetry-collector/32']
unit-opentelemetry-collector-32: 09:38:21 DEBUG unit.opentelemetry-collector/32.juju-log ======= Unit name: opentelemetry-collector/32, snap: node-exporter registered
unit-opentelemetry-collector-32: 09:38:21 DEBUG unit.opentelemetry-collector/32.juju-log ======= Unit name: opentelemetry-collector/32, snap: node-exporter used by: ['opentelemetry-collector/32']
```

We can see that the unit is registered with the snap.

Add the second subordinate relation:

```bash
$ juju integrate microsample-vm2 opentelemetry-collector
$ juju status
Model      Controller  Cloud/Region         Version  SLA          Timestamp
lxd-model  lxd         localhost/localhost  3.6.5    unsupported  09:40:20+08:00

App                      Version  Status  Scale  Charm                    Channel  Rev  Exposed  Message
microsample-vm1                   active      1  microsample-vm                      9  no
microsample-vm2                   active      1  microsample-vm                     10  no
opentelemetry-collector           active      2  opentelemetry-collector            16  no

Unit                           Workload  Agent  Machine  Public address  Ports  Message
microsample-vm1/1*             active    idle   4        10.15.210.174
  opentelemetry-collector/32*  active    idle            10.15.210.174
microsample-vm2/2*             active    idle   4        10.15.210.174
  opentelemetry-collector/33   active    idle            10.15.210.174

Machine  State    Address        Inst id        Base          AZ  Message
4        started  10.15.210.174  juju-8eb266-4  ubuntu@22.04      Running
```

Now there are two units of `opentelemetry-collector` running.

From `juju debug-log`, we can see that each snap is registered with two units:

```bash
unit-opentelemetry-collector-33: 09:40:04 DEBUG unit.opentelemetry-collector/33.juju-log ======= Unit name: opentelemetry-collector/33, snap: opentelemetry-collector registered
unit-opentelemetry-collector-33: 09:40:04 DEBUG unit.opentelemetry-collector/33.juju-log ======= Unit name: opentelemetry-collector/33, snap: opentelemetry-collector used by: ['opentelemetry-collector/32', 'opentelemetry-collector/33']
unit-opentelemetry-collector-33: 09:40:04 DEBUG unit.opentelemetry-collector/33.juju-log ======= Unit name: opentelemetry-collector/33, snap: node-exporter registered
unit-opentelemetry-collector-33: 09:40:04 DEBUG unit.opentelemetry-collector/33.juju-log ======= Unit name: opentelemetry-collector/33, snap: node-exporter used by: ['opentelemetry-collector/32', 'opentelemetry-collector/33']
```

If we remove the relation between the first principal charm (app `microsample-vm1`) and `opentelemetry-collector`, this is where the magic happens:

```bash
$ juju remove-relation microsample-vm1 opentelemetry-collector
$ juju status
Model      Controller  Cloud/Region         Version  SLA          Timestamp
lxd-model  lxd         localhost/localhost  3.6.5    unsupported  09:42:25+08:00

App                      Version  Status  Scale  Charm                    Channel  Rev  Exposed  Message
microsample-vm1                   active      1  microsample-vm                      9  no
microsample-vm2                   active      1  microsample-vm                     10  no
opentelemetry-collector           active      1  opentelemetry-collector            16  no

Unit                           Workload  Agent  Machine  Public address  Ports  Message
microsample-vm1/1*             active    idle   4        10.15.210.174
microsample-vm2/2*             active    idle   4        10.15.210.174
  opentelemetry-collector/33*  active    idle            10.15.210.174

Machine  State    Address        Inst id        Base          AZ  Message
4        started  10.15.210.174  juju-8eb266-4  ubuntu@22.04      Running
```

In the machine:

```bash
ubuntu@juju-8eb266-4:~$ ls /snap/
README  bin  core  core20  core24  lxd  microsample  node-exporter  opentelemetry-collector  snapd
```

We can see that the snaps aren't uninstalled, although the first subordinate relation is gone.

In `juju debug-log`:

```bash
unit-opentelemetry-collector-32: 09:42:12 DEBUG unit.opentelemetry-collector/32.juju-log ======= Unit name: opentelemetry-collector/32, snap: opentelemetry-collector used by: ['opentelemetry-collector/33']
unit-opentelemetry-collector-32: 09:42:12 DEBUG unit.opentelemetry-collector/32.juju-log ======= Unit name: opentelemetry-collector/32, snap: node-exporter used by: ['opentelemetry-collector/33']
```

The first unit is unregistered with these two snaps, but since the snaps are still registered with the second unit, meaning they are still being used, they won't be deleted.

Now, remove the second subordinate relation:

```bash
$ juju remove-relation microsample-vm2 opentelemetry-collector
$ juju status
Model      Controller  Cloud/Region         Version  SLA          Timestamp
lxd-model  lxd         localhost/localhost  3.6.5    unsupported  09:46:22+08:00

App                      Version  Status   Scale  Charm                    Channel  Rev  Exposed  Message
microsample-vm1                   active       1  microsample-vm                      9  no
microsample-vm2                   active       1  microsample-vm                     10  no
opentelemetry-collector           unknown      0  opentelemetry-collector            16  no

Unit                Workload  Agent  Machine  Public address  Ports  Message
microsample-vm1/1*  active    idle   4        10.15.210.174
microsample-vm2/2*  active    idle   4        10.15.210.174

Machine  State    Address        Inst id        Base          AZ  Message
4        started  10.15.210.174  juju-8eb266-4  ubuntu@22.04      Running
```

Both units of `opentelemetry-collector` are gone now.

In the machine:

```bash
$ ls /snap/
README  bin  core  core20  core24  lxd  microsample  snapd
```

Snaps are removed since there is no unit registered with them any more.

File: src/snap_lock.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+                CREATE TABLE IF NOT EXISTS config_locks (
+                    config_path TEXT PRIMARY KEY,
+                    locked_by TEXT,
+                    lock_time TIMESTAMP
+                )
+            """)
+            conn.commit()
+
+    def register(self, snap_name: str):
+        """Registers an instance using this snap."""
```
[the remark this answers is not in the document]
Comment (IronCore864) [reply src/snap_lock.py:57]: My intention for `lock_snap` is to acquire that lock first before snap install/remove. For register/unregister, it's OK not to acquire the lock. Register only means that the unit is now using the snap, increasing the reference counter.

**Comment by IronCore864** [comment]
Closing this PR in favor of https://github.com/canonical/opentelemetry-collector-operator/pull/6/files which implements the same feature using files/flock without the need of any extra database, even as lightweight as SQLite.
